### ISSUE TITLE
Native automatic mixed precision support

### ISSUE THREAD

**Issue by mcarilli** [pytorch/ignite#876]
Native automatic mixed precision support (`torch.cuda.amp`) is finally merged:
https://pytorch.org/docs/master/amp.html
https://pytorch.org/docs/master/notes/amp_examples.html

Apex Amp has many known pain points (extension builds, forward/backward compatibilty, DataParallel support, flaky checkpointing, i don’t even know if it can be hacked to handle double backward/gradient penalty, others…). `torch.cuda.amp` fixes all these, the interface is more flexible and intuitive, and the tighter integration with pytorch brings more future optimizations into scope.

I think the `torch.cuda.amp` API is a good fit for a higher-level library because its style is more functional (as in, it doesn't statefully alter anything outside itself). The necessary `torch.cuda.amp` calls don't have silent/weird effects elsewhere.

If you want to talk about adding `torch.cuda.amp` to Ignite, with an eye towards it becoming the future-proof source of mixed precision, message me on Pytorch slack anytime. I pinged you there as well but I’m not sure if you monitor it habitually.

**Comment by vfdev-5** [comment]
To close this issue, we can provide a colab notebook based on [Training Cycle-GAN on Horses to Zebras](https://github.com/pytorch/ignite/blob/master/examples/notebooks/CycleGAN.ipynb)

**Comment by mcarilli** [comment]
That would be great!  Did you already try it and it worked?

**Comment by vfdev-5** [comment]
- CycleGAN with AMP seems to have some problems.
Closes the issue as we provided a benchmark of AMP (Native, Apex) for Cifar-100: https://github.com/pytorch/ignite/blob/master/examples/notebooks/Cifar100_bench_amp.ipynb

**Comment by mcarilli** [comment]
Debugged the cyclegan example you sent, the problem appears unrelated to either Amp or your cyclegan script:  https://github.com/pytorch/pytorch/issues/37157

**Comment by vfdev-5** [comment]
As pytorch/pytorch#37157 was fixed in pytorch nightly we can retrain 
- apex based https://github.com/pytorch/ignite/blob/master/examples/notebooks/CycleGAN.ipynb
- torch native amp CycleGAN

**Comment by vfdev-5** [comment]
Online logging:
- https://app.wandb.ai/vfdev-5/ignite-cyclegan-torch-amp
  - https://colab.research.google.com/drive/1RRbI6ojMQUd7CYQ6Pio3IW5ZV35aguAP
- https://app.wandb.ai/vfdev-5/ignite-cyclegan-apex
  - https://colab.research.google.com/drive/1P2D_bYRn9l3I6t1veaKOsPWyxU-Y9-Jg
